src/lavinmq/auth/jwt.cr

```diff
@@ -0,0 +1,131 @@
+require "base64"
+require "json"
+require "openssl"
+require "./lib_crypto_ext"
+
+module JWT
+  class Error < Exception; end
+
+  class DecodeError < Error; end
+
+  class VerificationError < Error; end
+
+  class PasswordFormatError < Error; end
+
+  struct Token
+    getter header : JSON::Any
+    getter payload : JSON::Any
+    getter signature : Bytes
+
+    def initialize(@header, @payload, @signature)
+    end
+
+    def [](key : String)
+      @payload[key]
+    end
+
+    def []?(key : String)
+      @payload[key]?
+    end
+  end
+
+  class RS256Parser
+    def self.decode_header(token : String) : JSON::Any
+      parts = token.split('.')
+      raise DecodeError.new("Invalid JWT format: expected 3 parts") unless parts.size == 3
+      JSON.parse(base64url_decode(parts[0]))
+    end
+
+    def self.decode(token : String, public_key : String, verify : Bool = true) : Token
+      parts = token.split('.')
+      raise DecodeError.new("Invalid JWT format: expected 3 parts") unless parts.size == 3
+
+      header_json = base64url_decode(parts[0])
+      payload_json = base64url_decode(parts[1])
+      signature = base64url_decode_bytes(parts[2])
+
+      header = JSON.parse(header_json)
+      payload = JSON.parse(payload_json)
+
+      # Verify algorithm
+      alg = header["alg"]?.try(&.as_s)
+      raise DecodeError.new("Missing algorithm in header") unless alg
+      raise DecodeError.new("Expected RS256, got #{alg}") unless alg == "RS256"
+
+      # Verify signature if requested
+      if verify
+        signing_input = "#{parts[0]}.#{parts[1]}"
+        verify_signature(signing_input, signature, public_key)
+      end
+
+      Token.new(header, payload, signature)
+    end
+
+    def self.base64url_decode(str : String) : String
+      # Convert base64url to base64
+      base64 = str.tr("-_", "+/")
+
+      # Add padding if needed
+      case base64.size % 4
+      when 2
+        base64 += "=="
+      when 3
+        base64 += "="
+      end
+
+      Base64.decode_string(base64)
+    end
+
+    def self.base64url_decode_bytes(str : String) : Bytes
+      base64 = str.tr("-_", "+/")
+
+      case base64.size % 4
+      when 2
+        base64 += "=="
+      when 3
+        base64 += "="
+      end
+
+      Base64.decode(base64)
+    end
+
+    private def self.verify_signature(data : String, signature : Bytes, public_key_pem : String)
+      # Create a BIO with the PEM string
+      bio = LibCrypto.BIO_new(LibCrypto.bio_s_mem)
+      raise VerificationError.new("Failed to create BIO") if bio.null?
```

Comment:
To me that's not a `VerificationError`, that can only happen if we are out of memory, this could rather be a `RuntimeError`.